packages/bidi/init.lua

```diff
@@ -149,26 +149,33 @@ local splitNodelistIntoBidiRuns = function (typesetter)
   return nl
 end
 
-local bidiBoxupNodes = function (typesetter)
+local bidiBoxUpNodes = function (typesetter)
   local allDone = true
   for i = 1, #typesetter.state.nodes do
     if not typesetter.state.nodes[i].bidiDone then allDone = false end
   end
   if allDone then return typesetter:nobidi_boxUpNodes() end
-  local newNodeList = splitNodelistIntoBidiRuns(typesetter)
+  local newNodeList = splitNodelistIntoBidiRuns(nil, typesetter)
   typesetter:shapeAllNodes(newNodeList)
   typesetter.state.nodes = newNodeList
   local vboxlist = typesetter:nobidi_boxUpNodes()
   -- Scan for out-of-direction material
   for i = 1, #vboxlist do
     local v = vboxlist[i]
-    if v.is_vbox then package.reorder(nil, v, typesetter) end
+    if v.is_vbox then v.nodes = package.reorder(nil, v.nodes, typesetter) end
   end
   return vboxlist
 end
 
-function package.reorder (_, n, typesetter)
-  local nl = n.nodes
+-- We run UBA on hboxes here since they are not handled in bidiBoxUpNodes
+local bidiMakeHbox = function (typesetter, content)
+  local hbox, hlist = typesetter:nobidi_makeHbox(content)
+  hbox.value = splitNodelistIntoBidiRuns(hbox.value, typesetter)
+  hbox.value = package.reorder(nil, hbox.value, typesetter)
```

Comment:
Calling a package method with a nil self pointer and bypassing inheritance has heavy code smell.